src/action_install.rs

```diff
@@ -81,85 +86,62 @@ fn show_install_summary(pacman_deps: &IndexSet<String>, aur_packages: &IndexMap<
 	}
 }
 
-fn install_all(
-	dirs: &ProjectDirs,
-	split_to_depth: IndexMap<String, i32>,
-	split_to_pkgbase: IndexMap<String, String>,
-	offline: bool,
-	asdeps: bool,
-) {
-	let archive_whitelist = split_to_depth
-		.iter()
-		.map(|(split, _depth)| split.as_str())
-		.collect::<IndexSet<_>>();
-	trace!("All expected split packages: {:?}", archive_whitelist);
-	// get a list of (pkgbase, depth)
-	let packages = split_to_pkgbase.iter().map(|(split, pkgbase)| {
-		let depth = split_to_depth
-			.get(split)
-			.expect("Internal error: split package doesn't have recursive depth");
-		(pkgbase.to_string(), *depth, split.to_string())
-	});
-	// sort pairs in descending depth order
-	let packages = packages.sorted_by_key(|(_pkgbase, depth, _split)| -depth);
-	// Note that a pkgbase can appear at multiple depths because
-	// multiple split pkgnames can be at multiple depths.
-	// In this case, we only take the first occurrence of pkgbase,
-	// which would be the maximum depth because of sort order.
-	// We only take one occurrence because we want the package to only be built once.
-	let packages: Vec<(String, i32, String)> = packages
-		.unique_by(|(pkgbase, _depth, _split)| pkgbase.to_string())
+fn install_all(dirs: &ProjectDirs, actions: &Actions, as_deps: bool, offline: bool) {
+	let archive_whitelist = actions
+		.iter_build_pkgs()
+		.map(|pkg| pkg.pkg.name.as_str())
 		.collect::<Vec<_>>();
-	// once we have a collection of pkgname-s and their depth, proceed straightforwardly.
-	for (depth, packages) in &packages.iter().group_by(|(_pkgbase, depth, _split)| *depth) {
-		let packages = packages.collect::<Vec<&(String, i32, String)>>();
-		for (pkgbase, _depth, _split) in &packages {
-			let review_dir = rua_files::review_dir(dirs, pkgbase);
-			let build_dir = rua_files::build_dir(dirs, pkgbase);
-			rm_rf::force_remove_all(&build_dir).expect("Failed to remove old build dir");
-			std::fs::create_dir_all(&build_dir).expect("Failed to create build dir");
-			fs_extra::copy_items(
-				&vec![review_dir],
-				rua_files::global_build_dir(dirs),
-				&CopyOptions::new(),
+	trace!("All expected split packages: {:?}", archive_whitelist);
+
+	for base in &actions.build {
+		let pkgbase = &base.pkgbase;
+		let review_dir = rua_files::review_dir(dirs, pkgbase);
```

Comment:
Where is dependency grouping by depth now?
If we don't do that, then the user would be presented with "AUR repo review" for each package separately, with long pauses inbetween where package is actually built (see line 109 `wrapped::build_directory` in new code). If we go with this path, we have to give up the "minimizes user interactions" claim in README.md.

I don't want to do the latter much though, as you'd imagine, because I even wrote specific code myself to minimize it as much as theoretically possible. Even more agressive than grouping deps of individual AUR packages during recursion.

Reply:
I'm not sure how that would happen. This loop is all about reviewing. After the loop `install_all` is called.

So everything is reviewed first, then everything is build.

Reply:
Regarding separate loops: sorry, my own naming `rua_files::review_dir` confused me a bit. It can be read as a verb and looks very much alike `reviewing::review_repo`.

On the depth grouping though. You need it because you want to `tar_check` packages in a bulk. Or from another perspective, because the current PR does not have any grouping at all, you **do not install built tar archives until all packages are built**. This way, `makedepends` is not satisfied. Package installation with AUR `makedepends` will break.

Reply:
EDIT ^

Reply:
I guess you are right, in the sense that I tested it and it does fail because dependencies are not already installed. But I still don't really know what grouping is, or what tar_check really does.

I will however note that in the current system, it still seems to fail for the same reason. See pacaur for example.

```
morganamilo@Octavia ~ % rua install pacaur
Package pacaur depends on auracle-git. Resolving...

In order to install all targets, the following pacman packages will need to be installed:

And the following AUR packages will need to be built and installed:
  pacaur
  auracle-git

Proceed? [O]=ok, Ctrl-C=abort. o
Reviewing "/home/morganamilo/.config/rua/pkg/pacaur". 
[S]=run shellcheck on PKGBUILD, [D]=(identical to upstream, empty diff), [T]=run shell to edit/inspect, [O]=ok, use package m
Reviewing "/home/morganamilo/.config/rua/pkg/pacaur". 
[S]=run shellcheck on PKGBUILD, [D]=(identical to upstream, empty diff), [T]=run shell to edit/inspect, [O]=ok, use package o
Reviewing "/home/morganamilo/.config/rua/pkg/auracle-git". 
[S]=run shellcheck on PKGBUILD, [D]=(identical to upstream, empty diff), [T]=run shell to edit/inspect, [O]=ok, use package m
Reviewing "/home/morganamilo/.config/rua/pkg/auracle-git". 
[S]=run shellcheck on PKGBUILD, [D]=(identical to upstream, empty diff), [T]=run shell to edit/inspect, [O]=ok, use package o
==> Making package: pacaur 4.8.6-1 (Tue 10 Sep 2019 09:56:36 BST)
==> Checking runtime dependencies...
==> Missing dependencies:
  -> auracle-git
==> Checking buildtime dependencies...
==> ERROR: Could not resolve all dependencies.
Build failed with exit code 8 in /home/morganamilo/.cache/rua/build/pacaur
```

Reply:
Ah, interesting. That's a regression due to this issue: https://github.com/vn971/rua/issues/48

`makepkg` enforces not only makedepends to be installed when building, but also just `depends`. Since RUA will enforce makedepends to be present itself, this mechanism should be shut down via `makepkg --nodeps`. Too bad there's no separate option `makepkg --nomakedeps` in addition to `makepkg --nodeps`. I could then use the one that only disables direct dependencies.

I'll fix it today after work, just need to pass in `--nodeps`. The grouping and installation should be valid though (subject to testing after the fix).

Reply:
>makepkg enforces not only makedepends to be installed when building, but also just depends

The definition of depends are "things needed at runtime AND build time" while makedepends is "build time only". Passing -d is not really a fix because that will break pkgs that need the depends.

For example a package may call a command outside of any any functions. Thing setting some global variable  `_foo = $(command)`.

And also any git packages will fail to download their sources if git is not installed.